functions-continued/variables-are-references.md

```diff
@@ -0,0 +1,648 @@
+
+## Vocabulary and Synonyms
+
+| Vocab           | Definition                                                                                                                                                                                                               | Synonyms | How to Use in a Sentence |
+| --------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | -------- | ------------------------ |
+| Object Identity | An object's unique identity that never changes once it's been created. Usually a number. Similar to an object's address in memory. Find this with `id(obj)`. In a comparison that uses `is`, Python compares object IDs. | Object's address in memory | "I knew those two objects were different because they have different object IDs," "Maybe that variable's value is changing because two variables point to the same object with the same object ID"
+
+## Objects: Values, Types, and IDs
+
+Objects are a unit of data in programming that consist of three parts:
+
+1. Value
+2. Data type
+3. Identity
+
+"Object" is a fancy name for data "things" we've been working with so far.`"Hello, World!"` is an _object_, which has a value `"Hello, World!"` and a data type of `str`.
+
+More examples:
+
+- `33` is an object with a value `33` and a data type of `int`
+- `True` is an object with a value `True` and a data type of `bool`
+- `[81, 231, 12]` is an object with a data type of `list`. It contains three objects.
+
+An object's **identity** is an ID that a machine gives it when it's created. A machine will use an object's ID to find the correct object.
+
+While a program is running, every time an object is created, the object gets put into computer memory and is given an identity. We can think of this ID as the object's address in memory.
+
+We can see an object's ID by using the `id()` function.
+
+```python
+print('The ID of the string object "Hello, World!" is', id("Hello, World!"))
+```
+
+The `is` operator compares object IDs. **_An object's identity never changes once it has been created_**.
+
+## Variables Are References
+
+Variables are references that point to objects in memory.
+
+Consider this code:
+
+```python
+hello_world = "Hello, World!"
+current = 33
+is_enabled = True
+winners = [81, 231, 12]
+```
+
+- `hello_world` is a variable that refers to the object `"Hello, World!"` in memory
+- `current` is a variable that refers to the object `33` in memory
+- `is_enabled` is a variable that refers to the object `True` in memory
+- `winners` is a variable that refers to the object `[81, 231, 12]` in memory
```

Comment:
I'd love to make sure that I understand this well enough...

Does this wording still hold true (x out the "copy of" part)?

```suggestion
- `hello_world` is a variable that refers to the object `"Hello, World!"` in memory. It holds the object ID of `"Hello, World!"`.
- `current` is a variable that refers to the object `33` in memory. It holds the object ID of `33`.
- `is_enabled` is a variable that refers to the object `True` in memory. It holds the object ID of `True`.
- `winners` is a variable that refers to the object `[81, 231, 12]` in memory. It holds the object ID of `[81, 231, 12]`.
```

Reply:
It's still true. My intent in having the 'copy of' is to imply that this may not be the sole variable holding a reference to the object.  TBH, even saying that the variable holds the object ID is a little wishy washy, but I wanted to give a sense of what it means 'to refer' to an object.  i.e. the variable doesn't hold the object directly, but holds something that can be used to get to the object.

Reply:
What if...

```suggestion
- `hello_world` is a variable that refers to the object `"Hello, World!"` in memory. It holds something that can be used to get the object directly: a copy of the object ID of `"Hello, World!"`.
- `current` is a variable that refers to the object `33` in memory. It holds a copy of the object ID of `33`.
- `is_enabled` is a variable that refers to the object `True` in memory. It holds a copy of the object ID of `True`.
- `winners` is a variable that refers to the object `[81, 231, 12]` in memory. It holds a copy of the object ID of `[81, 231, 12]`.
```

Reply:
I'm starting to think this is trying to be very precise while still being imprecise.  Maybe just drop this whole line of thinking and just go with what you had originally.  Even with all the careful wording around object ID, of course, what the variables _really_ hold is a pointer to some PyObject "derived" type (which has the reference count, and the associated data-or pointer to data for more complicated types).

My main concern is whether students will have an intuitive sense of what 'refers to' means.  This would be  great spot for some visuals.  Chris' lists and references PR goes right for the jugular with memory addresses and diagrams.  Or maybe the less pedantic approach of the "additional analogies" (cut?) would help bridge that once we relax the idea insisting the analogies play true to the underlying implementation int their metaphor. I think I mentioned that in a comment I had put on them?

In any case, I'm leaning back towards reverting to your original version here.